File: java/src/test/java/it/giannibombelli/wsc2026/payment/application/usecases/PaymentExpiringTest.java
package it.giannibombelli.wsc2026.payment.application.usecases;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.ClientReference;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentDeadlineReached;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentEvent;
import it.giannibombelli.wsc2026.payment.domain.events.PaymentResultEvents;
import it.giannibombelli.wsc2026.payment.domain.payment.Payment;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentStatus;
import it.giannibombelli.wsc2026.payment.application.policies.PaymentExpiration;
import it.giannibombelli.wsc2026.payment.domain.ports.PaymentRepository;
import it.giannibombelli.wsc2026.payment.infrastructure.SqlitePaymentRepository;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import it.giannibombelli.wsc2026.testsupport.events.CapturingEventPublisher;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class PaymentExpiringTest {
    private PaymentRepository repository;
    private CapturingEventPublisher<PaymentEvent> publisher;
    private PaymentExpiring paymentExpiring;

    @BeforeEach
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeInMemoryDb("payment");
        repository = new SqlitePaymentRepository(dataSource);
        publisher = new CapturingEventPublisher<>();
        paymentExpiring = new PaymentExpiring(repository, publisher, new PaymentExpiration());
    }

    @Nested
    class Construction {
        @Test
        void rejectNullRepository() {
            assertThatThrownBy(() -> new PaymentExpiring(null, publisher, new PaymentExpiration()))
                .isInstanceOf(NullPointerException.class);
        }

        @Test
        void rejectNullPublisher() {
            assertThatThrownBy(() -> new PaymentExpiring(repository, null, new PaymentExpiration()))
                .isInstanceOf(NullPointerException.class);
        }

        @Test
        void rejectNullPolicy() {
            assertThatThrownBy(() -> new PaymentExpiring(repository, publisher, null))
                .isInstanceOf(NullPointerException.class);
        }
    }

    @Nested
    class Execution {
        @Test
        void rejectNullEvent() {
            assertThatThrownBy(() -> paymentExpiring.on(null))
                .isInstanceOf(IllegalArgumentException.class);
        }

        @Test
        void shouldExpirePayment() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            UUID clientReference = UUID.randomUUID();
            Money amount = new Money(new BigDecimal("50.00"));
            Instant requestedAt = Instant.parse("2026-06-07T10:00:00Z");
            Payment payment = Payment.request(paymentId, new ClientReference(clientReference), amount, new Timestamp(requestedAt));
            repository.save(payment);

            paymentExpiring.on(new PaymentDeadlineReached(paymentId));

            Optional<Payment> loaded = repository.findById(paymentId);
            assertThat(loaded).isPresent();
            assertThat(loaded.get().status()).isEqualTo(PaymentStatus.EXPIRED);
        }

        @Test
        void shouldPublishExpiredEvent() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);
            UUID clientReference = UUID.randomUUID();
            Money amount = new Money(new BigDecimal("50.00"));
            Instant requestedAt = Instant.parse("2026-06-07T10:00:00Z");
            Payment payment = Payment.request(paymentId, new ClientReference(clientReference), amount, new Timestamp(requestedAt));
            repository.save(payment);

            paymentExpiring.on(new PaymentDeadlineReached(paymentId));

            assertThat(publisher.events()).hasSize(1);
            assertThat(publisher.events().get(0)).isInstanceOf(PaymentResultEvents.PaymentExpired.class);
            PaymentResultEvents.PaymentExpired expired = (PaymentResultEvents.PaymentExpired) publisher.events().get(0);
            assertThat(expired.aggregateId()).isEqualTo(paymentId);
            assertThat(expired.clientReference().value()).isEqualTo(clientReference);
        }

        @Test
        void shouldFailIfPaymentNotFound() {
            PaymentId paymentId = EntityId.generate(PaymentId::new);

            assertThatThrownBy(() -> paymentExpiring.on(new PaymentDeadlineReached(paymentId)))
                .isInstanceOf(IllegalArgumentException.class);
        }
    }
}

File: typescript/tests/payment/application/usecases/paymentExpiring.test.ts
import { beforeEach, describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { ClientReference } from '@/common/domain/primitive/clientReference.js';
import { Money } from '@/common/domain/primitive/money.js';
import { Timestamp } from '@/common/domain/primitive/timestamp.js';
import { PaymentExpiring } from '@/payment/application/usecases/paymentExpiring.js';
import {
  paymentDeadlineReached,
  type PaymentDeadlineReached,
} from '@/payment/domain/events/paymentDeadlineReached.js';
import { PaymentEvent } from '@/payment/domain/events/paymentEvent.js';
import type { PaymentExpired } from '@/payment/domain/events/paymentResultEvents.js';
import { Payment } from '@/payment/domain/payment/payment.js';
import { PaymentId } from '@/payment/domain/payment/paymentId.js';
import { PaymentStatus } from '@/payment/domain/payment/paymentStatus.js';
import { PaymentExpiration } from '@/payment/application/policies/paymentExpiration.js';
import { SqlitePaymentRepository } from '@/payment/infrastructure/sqlitePaymentRepository.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { CapturingEventPublisher } from '../../../testsupport/events/capturingEventPublisher.js';
import { DatabaseSetup } from '../../../testsupport/databaseSetup.js';

describe('PaymentExpiring', () => {
  let repository: SqlitePaymentRepository;
  let publisher: CapturingEventPublisher<PaymentEvent>;
  let paymentExpiring: PaymentExpiring;

  beforeEach(() => {
    const database = DatabaseSetup.initializeInMemoryDb('payment');
    repository = new SqlitePaymentRepository(database);
    publisher = new CapturingEventPublisher();
    paymentExpiring = new PaymentExpiring(repository, publisher, new PaymentExpiration());
  });

  describe('construction', () => {
    it('rejects null repository', () => {
      expect(() => new PaymentExpiring(null as unknown as SqlitePaymentRepository, publisher, new PaymentExpiration())).toThrow();
    });

    it('rejects null publisher', () => {
      expect(() => new PaymentExpiring(repository, null as unknown as CapturingEventPublisher<PaymentEvent>, new PaymentExpiration())).toThrow();
    });

    it('rejects null policy', () => {
      expect(() => new PaymentExpiring(repository, publisher, null as unknown as PaymentExpiration)).toThrow();
    });
  });

  describe('execution', () => {
    it('rejects null event', () => {
      expect(() => paymentExpiring.on(null as unknown as PaymentDeadlineReached)).toThrow();
    });

    it('should expire payment', () => {
      const paymentId = seedPayment();

      paymentExpiring.on(paymentDeadlineReached(paymentId));

      const loaded = repository.findById(paymentId);
      expect(loaded).not.toBeNull();
      expect(loaded!.status()).toEqual(PaymentStatus.EXPIRED);
    });

    it('should publish expired event', () => {
      const paymentId = seedPayment();

      paymentExpiring.on(paymentDeadlineReached(paymentId));

      expect(publisher.events()).toHaveLength(1);
      expect(publisher.events()[0].kind).toBe('PaymentExpired');
      const expired = publisher.events()[0] as PaymentExpired;
      expect(expired.aggregateId).toEqual(paymentId);
    });

    it('should fail if payment not found', () => {
      const paymentId = generateId((value) => new PaymentId(value));

      expect(() => paymentExpiring.on(paymentDeadlineReached(paymentId))).toThrow();
    });
  });

  function seedPayment(): PaymentId {
    const paymentId = generateId((value) => new PaymentId(value));
    const payment = Payment.request(
      paymentId,
      new ClientReference(Uuid.fromString(crypto.randomUUID())),
      new Money(50),
      new Timestamp(new Date('2026-06-07T10:00:00.000Z'))
    );
    repository.save(payment);
    return paymentId;
  }
});
